python/pyspark/find_spark_home.py

```
@@ -0,0 +1,66 @@
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# This script attempt to determine the correct setting for SPARK_HOME given
+# that Spark may have been installed on the system with pip.
+
+from __future__ import print_function
+import os
+import sys
+
+
+def _find_spark_home():
+    """Find the SPARK_HOME."""
+    # If the enviroment has SPARK_HOME set trust it.
+    if "SPARK_HOME" in os.environ:
+        return os.environ["SPARK_HOME"]
+
+    def is_spark_home(path):
+        """Takes a path and returns true if the provided path could be a reasonable SPARK_HOME"""
+        return (os.path.isfile(os.path.join(path, "bin/spark-submit")) and
+                (os.path.isdir(os.path.join(path, "jars"))))
+
+    paths = ["../", os.path.join(os.path.dirname(sys.argv[0]), "../")]
+
+    # Add the path of the PySpark module if it exists
+    if sys.version < "3":
+        import imp
+        try:
+            paths.append(imp.find_module("pyspark")[1])
+        except ImportError:
+            # Not pip installed no worries
+            True
+    else:
+        from importlib.util import find_spec
+        try:
+            paths.append(os.path.dirname(find_spec("pyspark").origin))
+        except ImportError:
+            # Not pip installed no worries
+            True
+
+    # Normalize the paths
+    paths = map(lambda path: os.path.abspath(path), paths)
+
+    try:
+        return next(path for path in paths if is_spark_home(path))
+    except StopIteration:
+        print("Could not find valid SPARK_HOME while searching %s" % paths, file=sys.stderr)
```

We should raise an exception here or `exit(1)` since this is a fatal error.